### DISCUSSION TITLE
How to configure EngineCallbacks

### DISCUSSION THREAD

**Discussion by hkchenhongyi** [Novage/p2p-media-loader#432] [Q&A] [answered]
I want to get bandwidth 
https://novage.github.io/p2p-media-loader/docs/v2.0/types/p2p_media_loader_core.EngineCallbacks.html

onSuccess 

How to fill in the configuration information?

**Comment by DimaDemchenko** [comment]
@hkchenhongyi 
See [CoreEventMap](https://novage.github.io/p2p-media-loader/docs/v2.0/types/p2p_media_loader_core.CoreEventMap.html) for details on available events. For example, you can set up `onChunkDownloaded` event and calculate bandwidth by storing the `bytesLength` from event to measure download speed periodically.
```
config = {
        p2p: {
          core: {
            swarmId: "Optional custom swarm ID for stream",
          },
          onHlsJsCreated: (hls) => {
            hls.p2pEngine.addEventListener(
              "onChunkDownloaded",
              (bytesLength, downloadSource, peerId) => {
                // Sum the `bytesLength` of each chunk, and periodically
                // calculate bandwidth by dividing the total bytes by elapsed time.
              },
            );
          },
        },
      };

```

**Reply by hkchenhongyi** [reply]
> @hkchenhongyi 有关可用事件的详细信息，请参阅[CoreEventMap](https://novage.github.io/p2p-media-loader/docs/v2.0/types/p2p_media_loader_core.CoreEventMap.html)。例如，您可以设置`onChunkDownloaded`事件并通过存储`bytesLength`事件来计算带宽，以定期​​测量下载速度。
> 
> ```
> config = {
>         p2p: {
>           core: {
>             swarmId: "Optional custom swarm ID for stream",
>           },
>           onHlsJsCreated: (hls) => {
>             hls.p2pEngine.addEventListener(
>               "onChunkDownloaded",
>               (bytesLength, downloadSource, peerId) => {
>                 // Sum the `bytesLength` of each chunk, and periodically
>                 // calculate bandwidth by dividing the total bytes by elapsed time.
>               },
>             );
>           },
>         },
>       };
> ```

I see that the API provides
```
EngineCallbacks: {
onSuccess: ((response: SegmentResponse) => void);
onError: ((reason: CoreRequestError) => void);
}
```
But I don't know what the configuration writing format is, and I can't get the speed without the time you provided.

The API address I referenced is 
https://novage.github.io/p2p-media-loader/docs/v2.0/types/p2p_media_loader_core.EngineCallbacks.html

**Reply by DimaDemchenko** [reply, marked as answer]
You can refer to the Core package API if you're building a custom integration. We have Hls.js and Shaka packages built on the Core package, but they don’t provide direct access to engine callbacks. However, core events provide sufficient information to calculate bandwidth independently. One simple method is to sum all incoming chunks and use a 1-second interval to calculate bandwidth by dividing the total bytes by elapsed time, then resetting the total for the next interval

[Integration examples](https://novage.github.io/p2p-media-loader/docs/v2.0/index.html)
[Core API](https://novage.github.io/p2p-media-loader/docs/v2.0/modules/p2p_media_loader_core.html)
[Hls.js integration API](https://novage.github.io/p2p-media-loader/docs/v2.0/modules/p2p_media_loader_hlsjs.html)
[Shaka integration API](https://novage.github.io/p2p-media-loader/docs/v2.0/modules/p2p_media_loader_hlsjs.html)

```
let totalBytes = 0;
const interval = 1000;

const config = {
    p2p: {
        core: {
            swarmId: "Optional custom swarm ID for stream",
        },
        onHlsJsCreated: (hls) => {
            let totalBytes = 0; // Initialize the total bytes
            const interval = 1000; // Interval for bandwidth calculation in ms

            // Listen for chunk download events
            hls.p2pEngine.addEventListener("onChunkDownloaded", (bytesLength, downloadSource, peerId) => {
                totalBytes += bytesLength; 
            })
        },
    },
};

 setInterval(() => {
                const bandwidth = (totalBytes * 8) / interval; // Bandwidth in bits per second
                console.log(`Current bandwidth: ${bandwidth} bps`);
                
                totalBytes = 0; 
            }, interval);

```

This approach gives you an approximate bandwidth calculation by summing all chunks downloaded within each 1-second interval and resetting after every calculation

**Reply by hkchenhongyi** [reply]
> 如果您正在构建自定义集成，则可以参考核心包 API。我们在核心包上构建了 Hls.js 和 Shaka 包，但它们不提供对引擎回调的直接访问。但是，核心事件提供了足够的信息来独立计算带宽。一种简单的方法是将所有传入的块相加，并使用 1 秒的间隔来计算带宽，方法是将总字节数除以经过的时间，然后重置下一个间隔的总数
> 
> [集成示例](https://novage.github.io/p2p-media-loader/docs/v2.0/index.html) [核心 API ](https://novage.github.io/p2p-media-loader/docs/v2.0/modules/p2p_media_loader_core.html) [Hls.js 集成 API ](https://novage.github.io/p2p-media-loader/docs/v2.0/modules/p2p_media_loader_hlsjs.html) [Shaka 集成 API](https://novage.github.io/p2p-media-loader/docs/v2.0/modules/p2p_media_loader_hlsjs.html)
> 
> ```
> let totalBytes = 0;
> const interval = 1000;
> 
> const config = {
>     p2p: {
>         core: {
>             swarmId: "Optional custom swarm ID for stream",
>         },
>         onHlsJsCreated: (hls) => {
>             let totalBytes = 0; // Initialize the total bytes
>             const interval = 1000; // Interval for bandwidth calculation in ms
> 
>             // Listen for chunk download events
>             hls.p2pEngine.addEventListener("onChunkDownloaded", (bytesLength, downloadSource, peerId) => {
>                 totalBytes += bytesLength; 
>             })
>         },
>     },
> };
> 
>  setInterval(() => {
>                 const bandwidth = (totalBytes * 8) / interval; // Bandwidth in bits per second
>                 console.log(`Current bandwidth: ${bandwidth} bps`);
>                 
>                 totalBytes = 0; 
>             }, interval);
> ```
> 
> 此方法通过计算每 1 秒间隔内下载的所有块的总和并在每次计算后重置，为您提供近似的带宽计算结果

thank you very much！
